### PR TITLE
Suggest character encoding is incorrect when encountering random null bytes

### LINKED ISSUE #73979: Unhelpful Error Messages When Trying to Compile UTF16 Files
Saving a Hello World program as UTF-16:

```rust
fn main(){
    println!("Hello World!");
}
```
and trying to compile it causes about as many errors as there are characters in the file complaining of `unknown start of token: \u{0}`. Instead, using some heuristics to determine that the file is saved as UTF16 and printing a more helpful error message would be much friendlier to new users who are most likely to run into this issue.

### Meta

`rustc --version --verbose`:
```
rustc 1.46.0-nightly (50fc24d8a 2020-06-25)
binary: rustc
commit-hash: 50fc24d8a172a853b5dfe40702d6550e3b8562ba
commit-date: 2020-06-25
host: x86_64-unknown-linux-gnu
release: 1.46.0-nightly
LLVM version: 10.0
```

### PR DESCRIPTION
**Pull request by syvb**
This adds a note whenever null bytes are seen at the start of a token unexpectedly, since those tend to come from UTF-16 encoded files without a [BOM](https://en.wikipedia.org/wiki/Byte_order_mark) (if a UTF-16 BOM appears it won't be valid UTF-8, but if there is no BOM it be both valid UTF-16 and valid but garbled UTF-8). This approach was suggested in https://github.com/rust-lang/rust/issues/73979#issuecomment-653976451.

Closes #73979.

### CODE REVIEW

File: compiler/rustc_parse/src/lexer/mod.rs [hunk 1]
```diff
@@ -268,6 +268,9 @@ impl<'a> StringReader<'a> {
                 // tokens like `<<` from `rustc_lexer`, and then add fancier error recovery to it,
                 // as there will be less overall work to do this way.
                 let token = unicode_chars::check_for_substitution(self, start, c, &mut err);
+                if c == '\x00' {
+                    err.help("source files must be encoded in UTF-8, unexpected null bytes might occur when the wrong text encoding is used");
```
Comment (pickfire) [inline compiler/rustc_parse/src/lexer/mod.rs:272]: I was confused when I saw "wrong", maybe do we mean a "different" text encoding?

Comment (nagisa) [inline compiler/rustc_parse/src/lexer/mod.rs:272]: Wording nit: "The wrong text encoding" should read as "a wrong text encoding" or maybe "a different text encoding".

How about:

> source files must contain UTF-8 encoded text, unexpected null bytes might occur when a different encoding is used

Comment (nagisa) [reply compiler/rustc_parse/src/lexer/mod.rs:272]: Heh, pickfire was first.

**Pull request merged**